sql/2021/resource-hints/unused_preloaded_fonts.sql

```diff
@@ -0,0 +1,82 @@
+#standardSQL
+
+
+# returns the number of unused font files (assuming they are the same font based on the filename) that were preloaded
+# Ex: The below HTML will return 1: 
+#
+# <link rel="preload" href="./roboto.woff2" as="font" />
+# <link rel="preload" href="./roboto.woff" as="font" />
+# <link rel="preload" href="./montserrat.woff2" as="font" />
+CREATE TEMPORARY FUNCTION getUnusedFontDownloadsCount(almanac_string STRING)
+RETURNS INT64 LANGUAGE js AS ''' 
+try {
+  const almanac = JSON.parse(almanac_string);
+  if (Array.isArray(almanac) || typeof almanac != "object") return null;
+
+  let nodes = almanac["link-nodes"]["nodes"];
+  nodes = typeof nodes == "string" ? JSON.parse(nodes) : nodes;
+
+  const preloadedFonts = nodes.filter(
+    (n) => n["as"] === "font" && n["rel"] === "preload"
+  );
+
+  // extract the font file name and extension
+  const fontFiles = preloadedFonts.map((r) => {
+    const filePathParts = r.href.split("/");
+    return filePathParts[filePathParts.length - 1];
+  });
+
+  if (fontFiles.length === 0) {
+    return null;
+  }
+
+  // returns a key-value-pair with the filename and array of extensions used
+  // Ex: { "roboto": [".woff2", ".woff"], "montserrat": [".woff2"] }
+  const preloadedFontSet = fontFiles.reduce((acc, n) => {
+    const regexp = new RegExp("\\.[0-9a-z]+$");
+    const match = n.match(regexp);
+
+    if (match) {
+      const ext = match[0];
+      const filename = n.replace(ext, "");
+
+      return {
+        ...acc,
+        [filename]: [...(acc[filename] ? acc[filename] : []), ext],
+      };
+    }
+  }, {});
+
+  return Object.values(preloadedFontSet).reduce((acc, n) => {
+    if (n.find((m) => m === ".woff2")) {
```

Comment:
Any browser which supports `<link rel="preload" />` also supports .woff2; so preloading older font types (.woff, .eot, .ttf, .svg) for cross-browser compatibility when you have already preloaded woff2 is wasteful.

Having said that, this could be extended to .woff too.

Reply:
FYI we preload woff at work as don’t have woff2 versions. Must dig up that email thread as to why we can’t get them or if it’s possible to convert…

Anyway that’s one use case (maybe an edge one?) of why to preload older versions.

Reply:
In that case you'd just want to preload woff though, correct? I.e. you do not preload OTF, TTF, EOT...etc.

Reply:
Yes that's correct. Sorry didn't look at full context before butting in!

Reply:
Updated to include either woff and woff2.